galley/pkg/config/source/kube/apiserver/watcher.go

```diff
@@ -85,15 +85,12 @@ func (w *watcher) start() {
 	done := make(chan struct{})
 	w.done = done
 
-	// Send the FullSync event after the cache syncs.
-	go func() {
-		if cache.WaitForCacheSync(done, informer.HasSynced) {
-			w.handler.Handle(event.FullSyncFor(w.resource.Collection.Name))
-		}
-	}()
-
 	// Start CRD shared informer and wait for it to exit.
 	go informer.Run(done)
+	// Send the FullSync event after the cache syncs.
+	if cache.WaitForCacheSync(done, informer.HasSynced) {
+		go w.handler.Handle(event.FullSyncFor(w.resource.Collection.Name))
+	}
```

Comment:
Was this change also necessary to fix the race? It serializes the per-resource watch setup.

Reply:
Yes,  it has to wait cache sync before exit.